src/Database/Connection.php

```diff
@@ -653,22 +677,37 @@ public function transactional(callable $callback)
         try {
             $result = $callback($this);
         } catch (\Exception $e) {
-            $this->rollback();
+            $this->rollback(false);
             throw $e;
         }
 
         if ($result === false) {
-            $this->rollback();
+            $this->rollback(false);
 
             return false;
         }
 
-        $this->commit();
+        try {
+            $this->commit();
+        } catch (NestedTransactionRollbackException $e) {
+            $this->rollback(false);
+            throw $e;
+        }
 
         return $result;
     }
 
     /**
+     * Returns whether some nested transaction has been already rolled back.
+     *
+     * @return bool
+     */
+    public function wasNestedTransactionRolledback()
```

Comment:
Let's start with protected, if we need this method to prevent the call to those methods, then we make it public. I'd prefer hiding this implementation detail and see if there is a cleaner way of doing it.